apps/web/src/components/AssessmentStepTracker/utils.ts

```diff
@@ -3,17 +3,142 @@ import CheckCircleIcon from "@heroicons/react/20/solid/CheckCircleIcon";
+  const essentialAssessments = results.filter((result) => {
+    const containsEssentialSkill =
+      result.poolSkill?.type === PoolSkillType.Essential;
+    return (
+      containsEssentialSkill ||
+      result.assessmentResultType === AssessmentResultType.Education
+    );
+  });
+
+  const essentialWasUnsuccessful = !!essentialAssessments.find(
+    (result) => result.assessmentDecision === AssessmentDecision.Unsuccessful,
+  );
+  if (essentialWasUnsuccessful) return AssessmentDecision.Unsuccessful;
+
+  const containsOnHold = !!essentialAssessments.find(
+    (result) => result.assessmentDecision === AssessmentDecision.Hold,
+  );
+  if (containsOnHold) return AssessmentDecision.Hold;
+
+  const containsToAssess = !!results.find(
+    (result) => result.assessmentDecision === null,
+  );
+  if (containsToAssess) return NO_DECISION;
+
+  const essentialWasSuccessful = essentialAssessments.every(
+    (result) => result.assessmentDecision === AssessmentDecision.Successful,
+  );
+  if (essentialWasSuccessful) return AssessmentDecision.Successful;
+
+  return NO_DECISION;
+};
+
+type CandidateAssessmentResultNoDecision = Omit<
+  CandidateAssessmentResult,
+  "decision"
+>;
+
+/**
+ * Given a set of results, group them by candidate
+ * deriving the overall decision from all results
+ * for a specific candidate
+ *
+ * @param results
+ * @returns CandidateAssessmentResult[]
+ */
+export const groupResultsByCandidate = (
+  results: AssessmentResult[],
+): CandidateAssessmentResult[] => {
+  const groupedAssessments = results.reduce<
+    CandidateAssessmentResultNoDecision[]
+  >(
+    (
+      candidateResults: CandidateAssessmentResultNoDecision[],
+      currentResult: AssessmentResult,
+    ) => {
+      if (!currentResult.poolCandidate?.id) return candidateResults;
+      const candidateIndex = candidateResults.findIndex(
+        (candidateResult) =>
+          candidateResult.poolCandidate.id === currentResult.poolCandidate?.id,
+      );
+
+      // We store the candidate separate so not needed here
+      const strippedResult = omit(currentResult, "poolCandidate");
+
+      // Candidate does not exist in array so add it
+      if (candidateIndex === -1) {
+        return [
+          ...candidateResults,
+          {
+            poolCandidate: currentResult.poolCandidate,
+            results: [strippedResult],
+          },
+        ];
+      }
+
+      const foundCandidate = candidateResults[candidateIndex];
+      return [
+        ...candidateResults.slice(0, candidateIndex),
+        {
+          ...foundCandidate,
+          results: [...foundCandidate.results, strippedResult],
+        },
+        ...candidateResults.slice(candidateIndex + 1),
+      ];
+    },
+    [] as CandidateAssessmentResultNoDecision[],
+  );
+
+  return groupedAssessments.map((assessment) => ({
+    ...assessment,
+    decision: deriveResultsDecision(assessment.results),
+  }));
```

Comment:
This feels to me like it should be separate from the `groupResultsByCandidate` function.

Reply:
I started it out that way but quickly realized we just be calling it basically every where we use this group so that is why I included it in the return here. We use it in:

- Count on the summary of the step
- For the icon on the candidate
- For the sorting

So instead of calling it to get the decision in 3 different spots, it only needs to be called once here.